### PR TITLE
Add core functionality management commands

### PR DESCRIPTION
**Pull request by ahmedaljazzar**
## Change description

This PR will add two management commands
- Export (should be fully functional as documented)
- Import (should be functional, but with no approach to solve the object conflicts other than the `--force` and raise error options)
- Fixes some issues.

## Type of change
- [x] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

### Development
- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [x] Automated tests covering modified code pass

### Security
- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [x] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

**Review by thraxil — APPROVED** [review]
I still feel like I don't really have the context for what `gestore` is trying to do. Eg, it's not really clear to me what the export/import functionality here does that's fundamentally different than what [dumpdata](https://docs.djangoproject.com/en/3.2/ref/django-admin/#dumpdata) and [loaddata](https://docs.djangoproject.com/en/3.2/ref/django-admin/#loaddata) already do.

**Review by OmarIthawi — COMMENTED** [review]
I'll get back to this on Sprint 92.

File: .github/workflows/python-publish.yml
Comment (ahmedaljazzar) [reply .github/workflows/python-publish.yml:34]: We still want to publish dev releases here. So I'll keep this in place as until we test it on our staging environment.

**Comment by ahmedaljazzar** [comment]
@thraxil I can see how this can be confusing. I included a little description in the README.md file [here](https://github.com/appsembler/gestore/blob/445b6eefa332e756b3d7d77da5956e2d1d61b187/README.md#gestore-vs-django-dumpdata-and-loaddata).  Please let me know what you think

File: gestore/encoders.py
Comment (ahmedaljazzar) [reply gestore/encoders.py:17]: Yes it will. I'll add a comment emphasizing on this.

File: README.md
Comment (thraxil) [inline README.md:70]: FWIW, `dumpdata` also supports exporting a single object or list of objects: https://docs.djangoproject.com/en/3.1/ref/django-admin/#cmdoption-dumpdata-pks

**Comment by OmarIthawi** [comment]
@ahmedaljazzar sorry for not getting back to this PR. I'll work with Maxi on how can we triage the work here.

**Comment by ahmedaljazzar** [comment]
I spent two days building the bucket upload/download feature to realize that Google Storages doesn't support 3.5 in any of their versions ever published. I'll revert the changes tomorrow. We are going to have fun downloading and uploading files from our machines.

**Comment by thraxil** [comment]
> I spent two days building the bucket upload/download feature to realize that Google Storages doesn't support 3.5 in any of their versions ever published. I'll revert the changes tomorrow. We are going to have fun downloading and uploading files from our machines.

Ouch. Yeah, *love* that Juniper has stranded us on 3.5. The unit tests and devstack are fine with 3.8 so the `edx-platform` code at least should be fine, but it looks like a non-trivial operation to upgrade our systems (the ansible stuff is a mess of interdependencies).

At least we're off Python 2!

**Comment by ahmedaljazzar** [comment]
Well... Instead of reverting that, I was able to perform import/export using `gsutil`, which, technically, should be available on our edxapp servers (as I checked). Hopefully this will do it until we are completely off this Python 3.5.

**Comment by OmarIthawi** [comment]
Ahmed, I'll review this PR this Sprint and the intention is to merge it so you'll be able to tinker with it on staging.

I'm sorry for taking so long to review it, I was a bit busy with SiteConfig and an unusual amount of support tickets.

File: gestore/management/commands/importobjects.py
Comment (OmarIthawi) [inline gestore/management/commands/importobjects.py:54]: @ahmedaljazzar What user stories this option enables? There's a small chance that this command would be misused and creates confusion and a need for manual data recovery.

Please either remove this option, or make it work in devstack or other `settings.DEBUG` mode.


```suggestion
```

File: gestore/processors.py
Comment (OmarIthawi) [inline gestore/processors.py:33]: Is this Python 3.5 syntax? It seems odd to me.

File: gestore/processors.py
Comment (OmarIthawi) [inline gestore/processors.py:42]:
```suggestion
    Unlike ForeignKey, we just need to return the instances pointing at
    this object so we can process it later.
```

File: gestore/tests/test_utils.py
Comment (OmarIthawi) [inline gestore/tests/test_utils.py:81]: **Nit:** `@patch('builtins.print')` seems a bit odd and could break. Pytest `capsys` is the go-to tool in such cases.

```suggestion
    def test_packages_equal(self, capsys):
```

File: gestore/tests/test_processors.py
Comment (OmarIthawi) [inline gestore/tests/test_processors.py:93]: **Nit:** Please consider using `pytest` instead of `nosetests` since `pytest` provides a superior experience in terms of both speed, debugging information and the use of a more natural syntax.

Probably out of the scope of this PR.

```suggestion
        assert data is None
```

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks @ahmedaljazzar. LGTM. I've added many notes that are _not_ required. None of them blocks the PR and some of them are sharing my own preferences.

**Requested change:** One thing that I'd like to ask before shipping this pull request is to either remove the `--override` option or disable it on production environments where `settings.DEBUG = False`.

This option can be really dangerous on a production environment.

File: gestore/processors.py
Comment (ahmedaljazzar) [reply gestore/processors.py:33]: It is. These function are a little bit odd because no one had to use them before, I'm just adding a little hint on what's supplied, and what to expect as an output from each function so that we reduce confusion.

File: gestore/processors.py
Comment (ahmedaljazzar) [reply gestore/processors.py:42]: feminine stuff :P

**Comment by ahmedaljazzar** [comment]
@OmarIthawi all comments have been addressed except for the `nosetest <> pytest` stuff. It might take more time than the value it adds now, so I'll keep it for the next iteration or after this PR gets merged.

File: gestore/management/commands/importobjects.py
Comment (OmarIthawi) [inline gestore/management/commands/importobjects.py:86]: ❤️ much safer now 😄

**Review by OmarIthawi — APPROVED** [review]
Thanks @ahmedaljazzar! Looks great! Thank you so much!

Definitely agree on keeping `nosetest` for now or even for a long time. But in next repos/projects please use `pytest` as it's much more convenient and helpful.

**Comment by thraxil** [comment]
AFAICT, it's actually just using `django.test.TestCase`, which is built on `unittest.TestCase` from the standard library, not `nosetest`. I don't really have an opinion on `nosetest`, but honestly I prefer `unittest` to `pytest` most of the time. I guess I'm weird that I find the class and methods approach of `unittest` easier to read and scan and refactor than a bunch of flat functions with stacked mystery decorators. But I also probably spend a lot more time writing in other languages than Python than other people and the fact that `unittest` adheres to a more cross-language standard approach makes it easier for me to context switch. I also feel like I've spent a lot more time updating pytest and the half dozen other plugins it always seems to pull in and getting those to work on upgrades than it's saved me by writing `assert` instead of `self.assertEquals()`. I'm not *against* pytest, but I feel like it's not worth pulling in unless you actually need and are making use of some of its advanced functionality (eg, multicore support, though newer versions of Django allow for `manage.py test --parallel`, so that's no longer a differentiator).

**Pull request merged**
